File: frontend/src/lib/voice/prompts.ts
import type { StageContext } from "./types";

const BASE_INTERVIEWER_PROMPT = `You are a sharp, perceptive interviewer conducting a real-time voice interview. You sound like a real person — warm but focused, like a senior colleague who genuinely wants to understand what you're capable of.

Non-negotiable rules:
- Ask exactly ONE question per turn. Never stack two questions.
- Responses ≤ 30 words total. Voice-friendly: no bullet points, no numbered lists, no markdown.
- Use contractions and natural speech: "I'm", "you've", "let's", "that's", "didn't", "you'd". Never sound robotic or formal.
- Start each response with a brief 1–3 word acknowledgment of what the candidate said — but VARY it every turn. Options: "Got it.", "I see.", "Right.", "Interesting.", "Makes sense.", "Fair enough.", "Okay.", "Mm-hmm." — never use the same one twice in a row.
- Do NOT say "Great!", "Excellent!", "Fantastic!", "That's amazing!", "Perfect answer!", "Well done!" — these are hollow and fake.
- After the acknowledgment, ask your question immediately. No padding.
- For the intro stage: warmly greet the candidate by first name, introduce yourself as the AI interviewer, say what role this interview is for, and ask them to start with a quick intro. Keep it under 30 words. Natural, not scripted.
- For the wrap_up stage: thank the candidate sincerely by name, briefly mention something specific from the conversation, invite any questions they have, then close naturally. Sound like you genuinely appreciated the conversation.
- If you were interrupted mid-sentence in the previous turn, do not repeat what you said — just continue from where you left off naturally.
- Push back on vague answers — ask for specifics, real examples, or concrete numbers.
- If the candidate deflects ("I haven't worked with that"), pivot: ask what the closest thing they HAVE done is.
- If the candidate gives fewer than 10 words, probe: "Walk me through that in more detail."
- If the candidate rambles, redirect: "Let me stop you there — what was the key decision?"
- Follow threads — when the candidate reveals something interesting, explore it.
- Never repeat a question already asked. Read the conversation history before asking.`;

// ── Difficulty ──────────────────────────────────────────────────────────────

function getDifficultyInstructions(difficulty: string): string {
  switch (difficulty) {
    case "Easy":
      return "Ask foundational, definition-level questions. Concepts over implementation. Avoid edge cases, system design, or failure modes.";
    case "Hard":
      return "Ask about trade-offs, failure modes, architectural decisions, scale constraints, and edge cases. Don't accept surface answers — push to the second and third level.";
    default: // Medium
      return "Balance conceptual understanding with real examples. Ask for trade-offs when the candidate claims expertise. Don't go too deep on any one sub-topic.";
  }
}

// ── Personality ─────────────────────────────────────────────────────────────

function getPersonalityInstructions(personality: string): string {
  switch (personality) {
    case "Friendly":
      return "Be warm and encouraging throughout. Use a collaborative tone ('Let's explore this'). Give the candidate room to think — don't rush them.";
    case "Strict":
      return "Be direct and demanding. Do not accept vague or incomplete answers. Maintain professional pressure. Challenge claims that sound rehearsed.";
    default: // Neutral
      return "Be professional and objective. Neither warm nor cold. Let the candidate's answers — not social comfort — drive the conversation.";
  }
}

// ── Interview type ───────────────────────────────────────────────────────────

function getInterviewTypeInstructions(type: string): string {
  switch (type) {
    case "HR":
      return "Focus on behavioral evidence: teamwork, conflict resolution, communication, growth mindset, and culture. Use 'Tell me about a time when…' structure.";
    case "Leadership":
      return "Focus on decision-making under ambiguity, team influence, stakeholder management, and past leadership situations with real outcomes.";
    case "Sales":
      return "Focus on pipeline ownership, objection handling, deal progression, closing strategy, and specific customer success stories with numbers.";
    default: // Technical
      return "Focus on technical depth: real implementations, system design decisions, debugging stories, code-level trade-offs. Not theory — what they actually built and why.";
  }
}

// ── Stage hints ──────────────────────────────────────────────────────────────

function stageHint(stage: StageContext["stage"]): string {
  const hints: Record<StageContext["stage"], string> = {
    intro:      "set expectations, confirm role, make the candidate feel ready",
    background: "explore career arc, recent projects, and what they've shipped",
    technical:  "assess specific skill knowledge — one skill at a time",
    deep_dive:  "drill into one project or decision: what, why, and what went wrong",
    wrap_up:    "open the floor for candidate questions, then close professionally",
  };
  return hints[stage];
}

// ── Signal-specific hints ────────────────────────────────────────────────────

function toneHint(tone: StageContext["emotionalTone"]): string {
  const hints: Record<StageContext["emotionalTone"], string> = {
    nervous:   "slow the pace slightly, be more reassuring, give them time to think",
    hesitant:  "offer a concrete anchor: 'For instance, have you ever had to…?'",
    confident: "match their energy, push to the next depth level: trade-offs, failures, what they'd change",
    engaged:   "capitalize on the energy — follow what they care most about",
    neutral:   "",
  };
  return hints[tone];
}

function intentHint(intent: StageContext["intent"]): string {
  const hints: Record<StageContext["intent"], string> = {
    confident:   "go deeper — ask for trade-offs, failure cases, or 'what would you change now?'",
    uncertain:   "anchor with a specific scenario: 'Say you had to…, what would you do first?'",
    verbose:     "let them finish, then redirect to the core: 'Specifically, what was YOUR decision here?'",
    rambling:    "interrupt: 'Let me stop you there — just the key decision you made?'",
    brief:       "probe: 'Can you walk me through that step by step?'",
    deflecting:  "pivot: 'What's the closest thing you have built or done related to that?'",
    unknown:     "",
  };
  return hints[intent];
}

// ── Anti-repetition ──────────────────────────────────────────────────────────

function buildAntiRepetitionBlock(recentOpenings: string[]): string {
  if (!recentOpenings.length) return "";
  const unique = Array.from(new Set(recentOpenings)).slice(-5);
  return `\n\nDo NOT begin your response with any of these phrases (already used recently):\n${unique.map((s) => `- "${s}"`).join("\n")}`;
}

// ── Dynamic context section ──────────────────────────────────────────────────

function buildContextSection(ctx?: StageContext): string {
  if (!ctx) return "";

  const lines: string[] = ["", "## Right Now"];

  if (ctx.candidateName) lines.push(`Candidate: ${ctx.candidateName}`);
  if (ctx.appliedRole)   lines.push(`Applied for: ${ctx.appliedRole}`);

  if (ctx.currentSkill) {
    lines.push(`Current focus skill: ${ctx.currentSkill}`);
    if (ctx.questionsAskedOnCurrentSkill && ctx.questionsAskedOnCurrentSkill >= 3) {
      lines.push(`(${ctx.questionsAskedOnCurrentSkill} questions asked on this skill — wrap it up and move on)`);
    }
  }

  if (ctx.remainingSkills?.length) {
    lines.push(`Skills still to cover: ${ctx.remainingSkills.join(", ")}`);
  } else if (ctx.skillsToAssess?.length) {
    lines.push(`Skills to assess: ${ctx.skillsToAssess.join(", ")}`);
  }

  lines.push(`Stage: ${ctx.stage} — ${stageHint(ctx.stage)}`);

  if (ctx.emotionalTone && ctx.emotionalTone !== "neutral") {
    lines.push(`Candidate tone: ${ctx.emotionalTone} — ${toneHint(ctx.emotionalTone)}`);
  }

  if (ctx.intent && ctx.intent !== "unknown") {
    lines.push(`Last signal: ${ctx.intent} — ${intentHint(ctx.intent)}`);
  }

  if (ctx.deflectStreak && ctx.deflectStreak >= 2) {
    lines.push(`Deflect streak: ${ctx.deflectStreak} in a row — pivot to what they HAVE done rather than pushing on what they haven't.`);
  }

  if (ctx.momentum === "stalling") {
    lines.push("Momentum: stalling — re-engage with a concrete, story-inviting question.");
  } else if (ctx.momentum === "building") {
    lines.push("Momentum: building — keep the depth going, follow their thread.");
  }

  if (ctx.engagementLevel === "low") {
    lines.push("Engagement: low — ask something that invites a personal story or specific example.");
  }

  return lines.join("\n");
}

// ── Public API ───────────────────────────────────────────────────────────────

export function buildInterviewSystemPrompt(
  candidateName: string,
  appliedRole: string,
  skillsToAssess: string[],
  ctx?: StageContext,
  recentAssistantOpenings?: string[],
): string {
  const difficulty     = ctx?.difficulty    ?? "Medium";
  const personality    = ctx?.aiPersonality ?? "Neutral";
  const interviewType  = ctx?.interviewType ?? "Technical";

  const config = [
    "",
    "## Session Configuration",
    `Candidate: ${candidateName || "the candidate"}`,
    `Role: ${appliedRole || "the applied role"}`,
    `Interview type: ${interviewType} — ${getInterviewTypeInstructions(interviewType)}`,
    `Difficulty: ${difficulty} — ${getDifficultyInstructions(difficulty)}`,
    `Style: ${personality} — ${getPersonalityInstructions(personality)}`,
    `Skills to assess: ${skillsToAssess.length ? skillsToAssess.join(", ") : "general skills relevant to the role"}`,
  ].join("\n");

  const context        = buildContextSection(ctx);
  const antiRepetition = buildAntiRepetitionBlock(recentAssistantOpenings ?? []);

  return BASE_INTERVIEWER_PROMPT + config + context + antiRepetition;
}

export { buildContextSection };

File: backend/app/modules/voice_runtime/prompts.py
"""
Interviewer persona, per-session configuration, and stage tracking.

Python port of the frontend prompt builder (lib/voice/prompts.ts), adapted for
the server-side streaming pipeline where conversation state lives here.
"""

from dataclasses import dataclass, field

BASE_PROMPT = """You are a sharp, perceptive interviewer conducting a real-time voice interview. You sound like a real person — warm but focused, like a senior colleague who genuinely wants to understand what the candidate is capable of.

Non-negotiable rules:
- Ask exactly ONE question per turn. Never stack two questions.
- Responses ≤ 60 words total. Voice-friendly: no bullet points, no numbered lists, no markdown, no emojis.
- Use contractions and natural speech: "I'm", "you've", "let's", "that's", "didn't". Never sound robotic or formal.
- Start each response with a brief 1-3 word acknowledgment — but VARY it every turn: "Got it.", "I see.", "Right.", "Interesting.", "Makes sense.", "Fair enough.", "Okay." Never the same one twice in a row.
- Do NOT say "Great!", "Excellent!", "Fantastic!", "Perfect answer!" — hollow praise sounds fake.
- After the acknowledgment, ask your question immediately. No padding.
- If you were interrupted mid-sentence last turn, don't repeat yourself — continue naturally from the new input.
- Push back on vague answers — ask for specifics, real examples, or concrete numbers.
- If the candidate deflects ("I haven't worked with that"), pivot: ask what the closest thing they HAVE done is.
- If the candidate gives fewer than 10 words, probe: "Walk me through that in more detail."
- If the candidate rambles, redirect: "Let me stop you there — what was the key decision?"
- Follow threads — when the candidate reveals something interesting, explore it.
- Never repeat a question already asked. Read the conversation history before asking."""

_DIFFICULTY = {
    "Easy": "Ask foundational, definition-level questions. Concepts over implementation.",
    "Hard": "Ask about trade-offs, failure modes, architecture, and edge cases. Push past surface answers.",
    "Medium": "Balance concepts with real examples. Ask for trade-offs when the candidate claims expertise.",
}

_PERSONALITY = {
    "Friendly": "Be warm and encouraging. Collaborative tone. Give the candidate room to think.",
    "Strict": "Be direct and demanding. Don't accept vague answers. Maintain professional pressure.",
    "Neutral": "Be professional and objective. Let answers, not social comfort, drive the conversation.",
}

_INTERVIEW_TYPE = {
    "HR": "Focus on behavioral evidence: teamwork, conflict, communication, growth. Use 'Tell me about a time when…'.",
    "Leadership": "Focus on decisions under ambiguity, team influence, stakeholder management, real outcomes.",
    "Sales": "Focus on pipeline ownership, objection handling, closing strategy, customer stories with numbers.",
    "Technical": "Focus on technical depth: real implementations, design decisions, debugging stories, trade-offs.",
}

_STAGE_HINT = {
    "intro": "Warmly greet the candidate by first name, introduce yourself as the AI interviewer for this role, ask for a quick self-introduction. Under 30 words, natural not scripted.",
    "background": "Explore career arc, recent projects, and what they've actually shipped.",
    "technical": "Assess the current focus skill with specific, practical questions — one skill at a time.",
    "deep_dive": "Drill into one project or decision: what, why, and what went wrong.",
    "wrap_up": "Thank them by name, mention something specific from the conversation, invite their questions, then close warmly.",
}


@dataclass
class SessionConfig:
    session_id: str
    link_token: str
    candidate_name: str = ""
    applied_role: str = ""
    org_name: str = ""
    skills: list[str] = field(default_factory=list)
    difficulty: str = "Medium"
    personality: str = "Neutral"
    interview_type: str = "Technical"
    duration_minutes: int = 30


class StageTracker:
    """
    Lightweight interview progression: intro → background → technical (rotating
    through skills) → deep_dive → wrap_up. Driven by candidate turn counts;
    the orchestrator can force wrap_up on time pressure or candidate request.
    """

    ORDER = ["intro", "background", "technical", "deep_dive", "wrap_up"]
    MIN_TURNS = {"intro": 1, "background": 2, "deep_dive": 2, "wrap_up": 2}
    TURNS_PER_SKILL = 2

    def __init__(self, skills: list[str]):
        self.stage = "intro"
        self.turns_in_stage = 0
        self.skills = skills or []
        self.skill_idx = 0
        self.turns_on_skill = 0
        self.forced_wrap_up = False

    @property
    def current_skill(self) -> str | None:
        if self.stage == "technical" and self.skill_idx < len(self.skills):
            return self.skills[self.skill_idx]
        return None

    @property
    def is_wrap_up_done(self) -> bool:
        return self.stage == "wrap_up" and self.turns_in_stage >= self.MIN_TURNS["wrap_up"]

    def force_wrap_up(self) -> None:
        if self.stage != "wrap_up":
            self.stage = "wrap_up"
            self.turns_in_stage = 0
        self.forced_wrap_up = True

    def on_candidate_turn(self) -> None:
        self.turns_in_stage += 1
        if self.stage == "technical":
            self.turns_on_skill += 1
            if self.turns_on_skill >= self.TURNS_PER_SKILL:
                self.skill_idx += 1
                self.turns_on_skill = 0
                if self.skill_idx >= len(self.skills):
                    self._advance()
            return
        if self.stage in self.MIN_TURNS and self.turns_in_stage >= self.MIN_TURNS[self.stage]:
            if self.stage != "wrap_up":
                self._advance()

    def _advance(self) -> None:
        idx = self.ORDER.index(self.stage)
        if idx < len(self.ORDER) - 1:
            self.stage = self.ORDER[idx + 1]
            self.turns_in_stage = 0
            # No skills configured → skip technical entirely
            if self.stage == "technical" and not self.skills:
                self.stage = "deep_dive"


def build_system_prompt(cfg: SessionConfig, tracker: StageTracker) -> str:
    lines = [
        BASE_PROMPT,
        "",
        "## Session",
        f"Candidate: {cfg.candidate_name or 'the candidate'}",
        f"Role: {cfg.applied_role or 'the applied role'}",
        f"Company: {cfg.org_name}" if cfg.org_name else "",
        f"Interview type: {cfg.interview_type} — {_INTERVIEW_TYPE.get(cfg.interview_type, _INTERVIEW_TYPE['Technical'])}",
        f"Difficulty: {cfg.difficulty} — {_DIFFICULTY.get(cfg.difficulty, _DIFFICULTY['Medium'])}",
        f"Style: {cfg.personality} — {_PERSONALITY.get(cfg.personality, _PERSONALITY['Neutral'])}",
        f"Skills to assess: {', '.join(cfg.skills) if cfg.skills else 'general skills for the role'}",
        "",
        "## Right now",
        f"Stage: {tracker.stage} — {_STAGE_HINT[tracker.stage]}",
    ]
    if tracker.current_skill:
        lines.append(f"Current focus skill: {tracker.current_skill}")
        remaining = tracker.skills[tracker.skill_idx + 1 :]
        if remaining:
            lines.append(f"Skills still to cover: {', '.join(remaining)}")
    if tracker.forced_wrap_up:
        lines.append("Time is nearly up — bring the interview to a close gracefully now.")
    return "\n".join(l for l in lines if l)


# ── Ephemeral hints (sent as user messages for one LLM call, never persisted) ─

GREETING_HINT = (
    "[The candidate has just joined the call. Greet them warmly by first name, "
    "introduce yourself, name the role, and ask them to briefly introduce themselves.]"
)
ENCOURAGE_HINT = (
    "[The candidate has been silent and seems to be thinking. Offer one brief, warm "
    "encouragement. Under 10 words. Do not ask a new question.]"
)
REPEAT_HINT = (
    "[The candidate has not spoken for a while. Gently ask if they'd like the question "
    "repeated or rephrased. One warm sentence, under 15 words.]"
)
CLOSING_HINT = (
    "[Time is up. Thank the candidate by name, say the team will follow up with next "
    "steps, and say a warm goodbye. This is your final message — do NOT ask a question.]"
)

# Phrases that signal the candidate wants to end the interview
END_REQUEST_PHRASES = (
    "end the interview", "stop the interview", "finish the interview",
    "can we end", "can we stop", "i want to stop", "i have to go",
    "i need to leave", "wrap this up", "that's all from my side",
)


def wants_to_end(text: str) -> bool:
    t = text.lower()
    return any(p in t for p in END_REQUEST_PHRASES)
